### PR TITLE
Typescript Implementation

### PR DESCRIPTION
**Pull request by ahadrana**
This is a sample typescript implementation with some caveats. I tested the code but there is no test coverage yet for the complex parts of the code. 

The retry logic in createTransactionWithRetry might be overly complex, but it was not clear to me what the robustness requirements were or what the possible failure scenarios for the gateway are. So, my logic is as follows: (1) Make a request to gateway. (2) If this fails with a 500ish error, then assume transaction might be in flight, and do N getTransactionStatus calls to see if we can retrieve transaction status. If not, fail the request. (3) If TransactionStatus returns 404 (meaning transaction never made it across), recreate the transaction. Do up to M passes of the create/checkstatus loop with timeouts in between. 

Furthermore, after transaction create succeeds (via buster-api), there was some concern on my part that during my update of the DB record (with the buster transaction id) I might run into a race condition where the webhook triggers while I am doing the update. The concern being that I never want to update the DB record with a stale status. However, after thinking about it some more, it seems that during this particular database update, I don't actually update status (only external id). So, perhaps the extra level of redundancy I added in updateTransactionWithLastKnownStatus (which does an update with a where clause that contains a status of PENDING), is perhaps unnecessary. 

Lastly, getTransactions should probably limit the number of records it returns, to limit the potential for putting too much load on the API server and increasing latencies if the recordset in the DB is too large. 

Some additional minor notes: 

(1) I changed the /transactions POST endpoint to /transaction (singular)
(2) ngrok container in docker-compose file was exposing port 80 but default port exposed 
     by ngrok is 4040. Changed exposed port in ngrok to 4040 for now. 
(3) ngrok needs to be able to talk to js app container's web port. the env variable
     NGROK_PORT was pointing assumes localhost:NGROK_PORT. Unfortunately, the 
     the port is exposed via docker-compose host name (buster-js-app). Added links: 
     definition to buster-ngrok config, and changed NGROK_PORT to buster-js-app:4242. 
     Couldn't get NGROK_LOOK_DOMAIN to work properly. 
(4) I had to call docker port to find localhost mapping for ngrok port. Called 
     /api/listeners to discover external ngrok uri. 
(5) I Passed ngrok uri/webhooks to get-api-key.sh to get back API key (Google doc uses an
     old (invalid) ip instead of the buster.api hostname. 
(6)  Model.createTransaction returns a transaction object that has id null (which did not
     seem right, since id is an auto-incrementing pkey field. Looked at sequelize docs and 
     stackoverflow and added (primaryKey: true) to @Column definition for id column in 
     Transaction model defintion.
  
In conclusion, potentially this solution is overly complex in its error handling, but given the limited context I am working with, I felt I would put something out there that is _somewhat_ robust. In a real world scenario, I believe it is a slippery slope to try and accommodate lots of failure scenarios through complex conditional logic. However, it is all contextual. There could be scenarios where you need some level of retries and bulletproofing to minimize potential failures that can affect end users.

### CODE REVIEW

File: src/router.ts [hunk 1]
```diff
@@ -1,15 +1,140 @@
...
+        // validate that reference id matches
+        if (returnedRefId === undefined || returnedRefId !== referenceId) {
+            throw new Error('Buster TxnId:' + busterTxnId + ' returned invalid referenceId:' + returnedRefId + ' Expected:' + referenceId);
+        }
+        //now update transaction with buster id
+        //NOTE: we do two updates on the same transaciton, this one, unconditional, updates the
+        //transaction with the buster id. We don't want to lose this in case we never get a callback from buster
+        let txnOut:any = await model.updateTransaction(dbTxnId, { externalId: busterTxnId});
+        // Now, to prevent a race condition (where the webhook gets triggered prior to us updating the DB record
+        // we update the status field with a checked update 'where  status = PENDING'
```
Comment (ahadrana) [inline src/router.ts]: This is still not robust enough a check against overwriting status accidentally. The reality is that the transaction table should store last update timestamp (which comes from the buster api I believe). Then we can compare timestamp to timestamp returned from createTransaction call. A better way to guarantee consistency would be to read current record state in a transaction and then conditionally update the record with the same transaction. This way, a READ lock would be held on the record in this atomic context, between the read and the write. The webhook update code would need to do something similar.
